### DISCUSSION TITLE
Experimental CUDA Support

### DISCUSSION THREAD

**Discussion by MircoController** [thliebig/openEMS-Project#36] [Show and tell]
Over the holydays I had some time implementing a (rudimentary) GPU(CUDA) Support for openEMS. 
As this is a learning project for me I would like to share some findings and get feedback how to proceed further.

**Motivation:**

The main update functions for the E/H fields are heavily dependent on the system memory bandwidth but since *kindof accessible* GPUs have an order of magnitude higher bandwidth than comparable CPUs and HBM waiting around the corner, I gave it a try.

**Implementation:**
    
I started profiling the basic engine with valgrind and found the UpdateVoltage and UpdateCurrent functions to take the longest. The first goal was to port those functions and see what’s the maximum achievable speedup possible:

```

__global__ void UpdateVoltagesCUDAKernel(float* volt, float* curr, 
                                         const float* vv, const float* vi, 
                                         const unsigned int dimX, const unsigned int dimY, const unsigned int dimZ)
{
    unsigned int X = blockIdx.x*blockDim.x + threadIdx.x;
    unsigned int Y = blockIdx.y*blockDim.y + threadIdx.y;
    unsigned int Z = blockIdx.z*blockDim.z + threadIdx.z;

    // check boundaries
    if (X >= dimX) return;
    if (Y >= dimY) return;
    if (Z >= dimZ) return;

    // Calc the array index
    unsigned long n12 = dimY*dimX;
    unsigned long xyz = 3*(Z*n12 + Y*dimX + X);
    unsigned long xyzSZ = xyz - (3*n12)*(Z > 0);
    unsigned long xyzSY = xyz - (3*dimX)*(Y > 0);
    unsigned long xyzSX = xyz - 3*(X > 0);

    //do the updates here
    //for x
    volt[xyz+0] *= vv[xyz+0];
    volt[xyz+0] += vi[xyz+0] * (curr[xyz+2] - curr[xyzSY+2] - curr[xyz+1] + curr[xyzSZ+1]);

    //for y
    volt[xyz+1] *= vv[xyz+1];
    volt[xyz+1] += vi[xyz+1] * (curr[xyz+0] - curr[xyzSZ+0] - curr[xyz+2] + curr[xyzSX+2]);

    //for z
    volt[xyz+2] *= vv[xyz+2];
    volt[xyz+2] += vi[xyz+2] * (curr[xyz+1] - curr[xyzSX+1] - curr[xyz+0] + curr[xyzSY+0]);
}

void Engine_cuda::UpdateVoltages(unsigned int startX, unsigned int numX)
{
    dim3 block(32,1,1);
    dim3 grid((numLines[0]+block.x-1)/block.x, (numLines[1]+block.y-1)/block.y, (numLines[2]+block.z-1)/block.z);

    float* gpumem_volt = Get_N_3DArray_GPU(volt);
    float* gpumem_curr = Get_N_3DArray_GPU(curr);
    float* gpumem_vv = Get_N_3DArray_GPU(Op->vv);
    float* gpumem_vi = Get_N_3DArray_GPU(Op->vi);

    UpdateVoltagesCUDAKernel<<<grid, block>>>(gpumem_volt, gpumem_curr, gpumem_vv, gpumem_vi, numLines[0], numLines[1], numLines[2]);
}

```

As stated the kernel is pretty much dependent on the memory bandwidth of the system. To reduce the necessary memory accesses and benefit from the cacheline prefetching I copy the 4D Arrays to a 1D Array when moving the data to the GPU memory. This also works well since calculating the array index is basically free because of the immense computing power of the GPUs.

With this kernel (and a second for H-Field) running on GPU memory it will calculate **~2.5 GCells/s** compared to **~0.2 GCells/s** on CPU (single threaded).

A second limiting factor is the bandwidth between GPU and CPU so copy the data back and forth every timestep reduces the performance to an unusable way. Because of that i implemented the _Engine_cuda_ / _Operator_cuda_ in a way that the pre-processing will happen on the CPU and all arrays will be copied into the GPU before the first timestep. From that point on everything happens inside the GPU and will not copied back.
    
**Extensions:**
    
If everything happens inside the GPU, ideally the extensions should also run on the GPU. While occasionally accessing the GPU mem from CPU (GetVoltage/SetVoltage/..) will not influence the performance too much (e.g. for Voltage/Current Integral Processing) using it to calculate for example the Mur ABC will totally kill the performance. Therefore, I added specific functions, including CUDA kernels for the PML and excitation engine.

```
Extension(PML):

    case Engine::BASIC:
    ...
    case Engine::CUDA:
        {
            Engine_cuda* eng_cuda = (Engine_cuda*) m_Eng;
            eng_cuda->UPML_CUDA_PreUpdate(eng_cuda->get_volt(), m_Op_UPML->vv, m_Op_UPML->vvfo, volt_flux, m_Op_UPML->m_StartPos, m_Op_UPML->m_numLines);
            break;
        }   
    case Engine::SSE:
    ...


Engine_cuda:

    void Engine_cuda::UPML_CUDA_PreUpdate(float**** voltcurr, float**** vvii, float**** vviifo, float**** flux, unsigned int* pos_offset, unsigned int* upml_numLines) {

        // ... get managed GPU memory

        dim3 block(32,1,1);
        dim3 grid((upml_numLines[0]+block.x-1)/block.x, (upml_numLines[1]+block.y-1)/block.y, (upml_numLines[2]+block.z-1)/block.z);

        // run the CUDA kernel
        UPML_Kernel_PreUpdate<<<grid, block>>>(voltcurr_gpumem,
                                               flux_gpumem,
                                               vvii_gpumem,
                                               vviifo_gpumem,
                                               pos_offset[0], pos_offset[1], pos_offset[2], 
                                               upml_numLines[0], upml_numLines[1], upml_numLines[2],  
                                               numLines[0], numLines[1]); 
    }   
```

Since those functions differ that much from the Basic and SSE implementation and require some CUDA specific headers, I put all CUDA specific functions inside the _Engine_cuda_ class, to keep the extensions clean and readable. However, this blows up the _Engine_cuda_ class. Do you have any other thoughts/ideas how to deal with this?
    
**Results:**

I initial got into touch with openEMS searching for a fieldsolver for simulation distributed element filters. So I my experiments are based on the hyp2mat examples especially the hairpin filter. To increase the number of points, i simply divided the "resolution" variable by /10 and /20.

```
Hairpinfilter (hyp2mat, PML on all sides)
Model size                 mt+sse      CUDA         speedup
78x49x26    (~0.1M Cells)  114.4 MC/s   447.4 MC/s  3,91
241x229x89  (~5M Cells)    138.5 MC/s  1154.6 MC/s  8,34
463x440x161 (~33M Cells)   192.6 MC/s  1429.5 MC/s  7,42
```

_Setup:_
_openEMS 64bit -- version v0.0.35_
_AMD Ryzen 5 5600X 6-Core Processor_
_NVIDIA GeForce RTX 3060_

_(Hopefully there are no stupid mistakes anywhere 😊)_ 

**Next possible steps:**
    
- With some design/coding guidance it should be possible to get a first working CUDA version going, despite the fact that some extensions aren’t fully supported / work not very performant.

- Currently multithreading for the "pre-processing" isn't supported, I haven’t found an easy way to get it going without changing the code too much. Any ideas on that?

- It should be possible to support multi-GPU processing, but I don’t know how high the penalty will be when synchronizing the boundaries of the divided grid between the GPUs. I might give it a try.

- For calculation intensive extensions like PML it might be a good idea to incorporate the calculation directly into the main UpdateVoltage/Current CUDA kernel. I roughly tried it for PML and got a speedup of **~1.4x** (1154.6 MC/s -> 1655.1 MC/s). But this works against the idea of having separated extensions. What’s your thoughts on that?

- The current CUDA kernels are very simple but there are smarter ways to write a FDTD kernel which do not rely on the cache but use shared memory to reuse data more frequently. I gave it a try but the kernel is much more complex and the speedup wasn’t that big ~10%, so I ditched the idea for now.

**Comment by biergaizi** [comment]
> For calculation intensive extensions like PML it might be a good idea to incorporate the calculation directly into the main UpdateVoltage/Current CUDA kernel. I roughly tried it for PML and got a speedup of ~1.4x (1154.6 MC/s -> 1655.1 MC/s). But this works against the idea of having separated extensions. What’s your thoughts on that?

I recently also noticed this issue in the CPU kernel as well, and made a proposal at https://github.com/thliebig/openEMS/issues/100. 

To repeat it here:

Now the problem is whether it can be optimized without sacrificing modularity. I'm now seeing two potential solutions.

First is to use (single-thread) domain decomposition. Instead of going through the entire 3D space at a time, perhaps the update can be splitted into multiple blocks. Instead of running pre-update, main-update and post-update across the entire 3D space at once, as in:

```
for (int x = 0; x < x_max; x++) {
    for (int y = 0; y < y_max; y++) {
        for (int z = 0; z < z_max; z++) {
            pre_update(x, y, z)
        }
    }
}

for (int x = 0; x < x_max; x++) {
    for (int y = 0; y < y_max; y++) {
        for (int z = 0; z < z_max; z++) {
            fdtd_kernel(x, y, z)
        }
    }
}

for (int x = 0; x < x_max; x++) {
    for (int y = 0; y < y_max; y++) {
        for (int z = 0; z < z_max; z++) {
            post_update(x, y, z)
        }
    }
}
```

It can perhaps be performed in a piecewise manner, allowing the main FDTD kernel and the plugins can reuse the data in CPU cache.

```
int x_start[], x_end[];
int y_start[], y_end[];
int z_start[], z_end[];
int domain_len;

for (int domain = 0; domain < domain_len; domain++) {
    for (int x = x_start[domain] x < x_end[domain]; x++) {
        for (int y = y_start[domain]; y < y_end[domain]; y++) {
            for (int z = z_start[domain]; z < z_end[domain]; z++) {
                pre_update(x, y, z)
            }
        }
    }

    for (int x = x_start[domain] x < x_end[domain]; x++) {
        for (int y = y_start[domain]; y < y_end[domain]; y++) {
            for (int z = z_start[domain]; z < z_end[domain]; z++) {
                fdtd_kernel(x, y, z)
            }
        }
    }

    for (int x = x_start[domain] x < x_end[domain]; x++) {
        for (int y = y_start[domain]; y < y_end[domain]; y++) {
            for (int z = z_start[domain]; z < z_end[domain]; z++) {
                post_update(x, y, z)
            }
        }
    }
}
```

Unfortunately, if I read the code correctly, full 3D domain decomposition is not supported by openEMS. It only supports decompositing the domain across the X axis. So if there are many cells at the Y and Z directions, it still won't fit in cache. Implementing this solution needs adding full 3D partitioning to all update functions.

Another possible solution is allowing a plugin to insert itself dynamically into main field update loop of the main kernel to "fuse" the computation together, rather than using its own redundant loop. This way a few memory accesses can be saved.

```
for (int x = 0; x < x_max; x++) {
    for (int y = 0; y < y_max; y++) {
        for (int z = 0; z < z_max; z++) {
            if (has_extension(x, y, z)) {
                pre_update(x, y, z);
                fdtd_kernel(x, y, z);
                post_update(x, y, z);
            }
            else {
                fdtd_kernel(x, y, z)
            }
        }
    }
}
```

Modularity can still be somewhat preserved by using function pointers instead of hardcoding extension into the main engine.

Both solutions also have the problem of losing the ability to randomly access a coherent electromagnetic field at any point in space from a plugin, so it cannot be applied to all extensions.

However, it can perhaps at least be a fast path for plugins that don't need access to the whole field, like PML.

**Reply by thliebig** [reply]
Please correct me if I'm wrong, but calling functions inside the inner loop is a very bad idea?? At least if the function can not be inlined?

**Reply by biergaizi** [reply]
> Please correct me if I'm wrong, but calling functions inside the inner loop is a very bad idea?? At least if the function can not be inlined?

Yes, I expect an overhead. Also, determining whether a cell has an associated extension by an `if` statement should also be bad for performance, since it introduces branching. My hope for is that, despite these two sources of overheads, the final result is still a net speedup due to less data movements and pressure on memory.

If this is not possible, I also have an idea of implementing this using an extreme method - rewrite simulation kernels and extensions as lightweight "scripts", then during the initialization of simulation, a machine-code kernel is generated on-the-fly, specifically inlined depending on the extensions used and locations of the cells for a particular simulation. With high-level libraries like GCC-JIT, it should be possible to write a machine code generator without too much difficulty. For now, I still plan to try the first method, the JIT method is just a possible idea that remains on paper.

**Comment by JohannesLau** [comment]
Excellent work so far, the potential speed up looks very promising. 
Are you interested in any help benchmarking different cases or on different hardware, or just general testing, if so I would be happy to help.

**Comment by biergaizi** [comment]
I'm also experimenting with GPU computing. Instead of CUDA, I'm using SYCL. SYCL is a vendor-neutral, high-level GPU computing standard by Khronos Group, just like OpenCL. But the differences are:

1. Operating at a higher abstraction level than OpenCL.
2. SYCL 2020 is based on C++17, not plain C like OpenCL, so it's more suited for a C++ codebase.
3. Multiple compilers / implementations exist. Some compilers can target Nvidia CUDA and AMD HIP directly without OpenCL. So unlike the incomplete OpenCL support by Nvidia, SYCL's performance is close to native on Nvidia hardware, within 20%.
4. Intel is heavily promoting it as "oneAPI" for their GPUs, but it's just a marketing name for Intel's development toolchain based on SYCL. SYCL itself is vendor-neutral.

So I see it as a potential pathway for porting FDTD to Nvidia, AMD and Intel GPUs.

The initial result: without any optimization, for a plain FDTD kernel with only a E-field and H-field update loop, simulation speed is 1.0 Gcells/s for a 200x200x200 grid on a AMD RX580 GPU with 256 GB/s memory bandwidth.

**Comment by biergaizi** [comment]
After applying the most basic vectorization, simulation speed increased from 1.0 Gcells/s to 1.3 Gcells/s. According to my calculation, this speed corresponds to a DRAM bandwidth efficiency of 75% (of RX580's peak 256 GB/s). So in theory, another 25% speedup to 1.6 Gcells/s is perhaps possible. I also just ordered an AMD Radeon Instinct MI50 GPU (its desktop version is known as Radeon VII) for further experiments. These used MI50 can be purchased for $100 to $150 these days (*), they're equipped with 4096-bit HBM2 memory running at 1024 GB/s. They seem to be an ideal candidate for running FDTD simulations, allowing a simulation speed around 5 Gcells/s (projected speedup, not tested yet).

But all of these experiments are only "laboratory prototypes" running with nothing but a FDTD kernel main loop. To make real speedup in actual openEMS software, several obstacles must be overcome...

1. The field dump problem. openEMS performs field dump at the end of every timestep, the field dump will create extra memory traffic, in addition, communication latency will completely stop the simulation at every timestep. The latency problem should be solvable - just use GPU VRAM to buffer field dump and transfer them to the CPU after 100 timesteps. But the memory bandwidth is still a problem, especially when "multi-timestep" optimization techniques are used, which makes it a problem even on the CPU - if it's the case, it would be impossible to obtain field dump at every timestep. Is it possible to reduce the frequency of the field dump without affecting near-field to far-field transformation? (CC @thliebig)

2. Extensions. As an academic project, the openEMS architecture is not designed for maximum speed but modularity, almost all functionalities beyond the main FDTD update loop is implemented as extensions, including excitation source, PML, conducting sheets, etc. These extensions will create a lot of additional memory traffic, so the actual speed will be much lower than 1 GC/s or 5 GC/s. I still hope to eventually solve this problem using the method I already described above. 

(*) Apparently these are not actual MI50, but a custom stripped-down version of desktop Radeon VII, and sold as Instinct MI50, specifically for coin mining operations. Obviously now they became worthless and start circulating around the second-hand markets in China. Radeon VII is extremely slow at doing FP64, but for openEMS, all it matters is FP32.

**Reply by thliebig** [reply]
As for 2) I think for maybe this high performance engines it would be okay to implement some of the extensions directly into the main (mutlistepping) FDTD? The operators could still be used to calculate the required coefficients, but maybe the engine extensions would be unused?

**Comment by biergaizi** [comment]
> Is it possible to reduce the frequency of the field dump without affecting near-field to far-field transformation? (CC @thliebig)

I think I've already found the answer. In openEMS, the field dump is *not* performed at every timestep, but instead is processed in batch.

        int step=PA->Process();
        if ((step<0) || (step>(int)NrTS)) step=NrTS;
        while ((FDTD_Eng->GetNumberOfTimesteps()<NrTS) && (change>endCrit) && !CheckAbortCond())
        {
                FDTD_Eng->IterateTS(step);
                step=PA->Process();
                // [...]
        }

`step` is typically hundreds of timesteps. So the main engine loop should already be a natural fit for multi-timestep simulation, which is great news.

**Reply by thliebig** [reply]
Yes, there is a stepping interval at which all fields have to be up to date, but that step is dependent on the (FDTD) timestep and max. requested frequency (etc.) and is usually big enough for what you want to do. And no dump or energy estimation or anything like that should access the fields in between these steps.
But any multiple time-stepping should try to somehow fit into that "larger" stepping scheme... 
But yes, I see no reason why multiple stepping should not work in openEMS...

**Comment by biergaizi** [comment]
Using the AMD Radeon VII / Instinct MI50 GPU, under fully ideal conditions, my main FDTD test kernel has just set a new speed record of 6100 MC/s!

Note that all the previous disclaimers still apply, it's only a prototype, prepare for perhaps half the speed in a real simulation...
